# Part 1. Equation of a Slime

How many late days are you using for this assignment?
Zero

In [1]:
# Imports section
import pandas as pd

## 1. Loading the dataset

In [3]:
# Using pandas, load the dataset
data = pd.read_csv('science_data_large.csv')

# Output the first 15 rows of the data
print(data.head(16))

# Display a summary of the table information (number of datapoints, etc.)
print(data.info())


    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
15              62       293  3.952883e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non

## 2. Splitting the dataset

In [9]:
from sklearn.model_selection import train_test_split

# Take the pandas dataset and split it into our features (X) and label (y)
X = data[['Temperature °C', 'Mols KCL']]  # Adjusted feature names
y = data['Size nm^3']  # Adjusted label name

# Use sklearn to split the features and labels into a training/test set (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)


## 3. Perform a Linear Regression

In [22]:
from sklearn.linear_model import LinearRegression
import numpy as np

# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(X_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
sample_data = pd.DataFrame(np.array([[369, 112]]), columns=['Temperature °C', 'Mols KCL'])  
predicted_size_change = model.predict(sample_data)
print(f'Predicted size change: {predicted_size_change[0]}')

# Report the score for that model using the default score function property of the SKlearn model
score = model.score(X_test, y_test)
print(f'Model score (R^2): {score}')



# Extract the coefficients and intercept from the model
coefficients = model.coef_
intercept = model.intercept_

print(f'Coefficient for Temperature °C (m1): {coefficients[0]}')
print(f'Coefficient for Mols KCL (m2): {coefficients[1]}')
print(f'Intercept (b): {intercept}')



Predicted size change: 25878.394397240365
Model score (R^2): 0.8552472077276095
Coefficient for Temperature °C (m1): 866.1464133719209
Coefficient for Mols KCL (m2): 1032.6950664857964
Intercept (b): -409391.47958340764


In your own words (markdown), explain what the score means:
The score represents the coefficient of determination (R²), which indicates how well the independent variables (Temperature and Mols KCL) explain the variance in the dependent variable (Size nm^3). A score of 1.0 indicates a perfect fit, while a score of 0 indicates that the model does not explain any variability.

Write the linear equation of a slime: (example equation: $E = mc^2$)


$$
h(x) = 866.14641 \cdot \text{Temperature °C} + 1032.69507 \cdot \text{Mols KCL} - 409391.47958
$$

Report on score and explain meaning:
### Linear Equation of a Slime

####  Values:
- Coefficient for Temperature °C (m1): \( 866.14641 \)
- Coefficient for Mols KCL (m2): \( 1032.69507 \)
- Intercept (b): \( -409391.47958 \)

### Report on Model Score

#### Model Score (R²)
- **Value**: \( R^2 = 0.85525 \)

**Explanation**:
- The R² value (Coefficient of Determination) quantifies how well the independent variables (Temperature °C and Mols KCL) explain the variance in the dependent variable (size change).
- An R² score of \( 0.85525 \) indicates that approximately \( 85.52\% \) of the variance in size change can be explained by the linear regression model.
- This high score suggests a strong relationship between the temperature and KCl concentration with the size change, indicating that these variables are significant predictors in the model. However, it also implies that about \( 14.48\% \) of the variance may be attributable to other factors not included in the model or random noise.

The model score of \( R^2 = 0.85525 \) indicates a robust predictive capability, suggesting that both temperature and KCl concentration are significant factors influencing the growth of the slime, with room for improvement to capture other influential aspects that may not be accounted for in the current model.


## 4. Use Cross Validation

In [31]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42

# Report on their finding and their significance

from sklearn.model_selection import KFold, cross_val_score

# Setup KFold cross-validator with a fixed random_state
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Use cross_val_score with the KFold object
cross_val_scores = cross_val_score(model, X, y, cv=kf)
print(f'Cross-validation scores: {cross_val_scores}')
print(f'Mean cross-validation score: {cross_val_scores.mean()}')


Cross-validation scores: [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
Mean cross-validation score: 0.8597294202684646


Write findings here:

Cross-Validation Scores

The results from the cross-validation process yielded the following R² scores for each fold:
Cross-validation scores:
- Fold 1: 0.86152
- Fold 2: 0.82742
- Fold 3: 0.87195
- Fold 4: 0.88166
- Fold 5: 0.85609

Mean Cross-Validation Score

The mean cross-validation score is calculated as:

Mean R²: 0.85973

Interpretation and Significance

High R² Values:

- The individual R² scores range from approximately 0.82742 to 0.88166, indicating that the model explains roughly 82.742% to 88.166% of the variance in the size change (the dependent variable) across different subsets of the data. These values are relatively high and suggest that the model is capturing a substantial portion of the variability associated with the relationship between the independent variables (Temperature °C and Mols KCL) and the dependent variable.

Consistency Across Folds:

- The variability among the fold scores indicates a strong and consistent model performance, although there is a noticeable dip in the second fold (0.82742) compared to others. This could suggest that this particular fold may contain some data characteristics (like outliers or a less representative sample of the overall dataset) that make it more challenging for the model to predict accurately. However, the other folds compensate for this, with R² values still remaining relatively high.

Mean Score:

- The mean R² value of 0.85973 indicates that, on average, the model is performing well in predicting the size change when exposed to new, unseen data. This score suggests that you can expect the model to generalize relatively well beyond the training data.

Practical Implications:

- The findings imply that the combination of Temperature and Mols KCL is a robust predictor of the size of the slime. Given a new combination of these factors, the model should provide a reliable estimate of the size change based on its training from the dataset.

## 5. Using Polynomial Regression

In [35]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)

# Report on the metrics and output the resultant equation as you did in Part 3.

from sklearn.preprocessing import PolynomialFeatures
from sklearn.model_selection import KFold, cross_val_score
from sklearn.linear_model import LinearRegression

# Using the PolynomialFeatures library to perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(degree=2)
X_poly = poly.fit_transform(X)

# Setup KFold cross-validator
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Train a polynomial regression model
poly_model = LinearRegression()

# Perform k-fold cross validation
poly_cross_val_scores = cross_val_score(poly_model, X_poly, y, cv=kf)
print(f'Polynomial cross-validation scores: {poly_cross_val_scores}')
print(f'Mean polynomial cross-validation score: {poly_cross_val_scores.mean()}')

# Train the model on the entire dataset after k-fold validation
poly_model.fit(X_poly, y)

# Report on the metrics and output the resultant equation
poly_coefficients = poly_model.coef_
poly_intercept = poly_model.intercept_

# Output the coefficients and intercept
print(f'Polynomial Coefficients: {poly_coefficients}')
print(f'Polynomial Intercept: {poly_intercept}')

Polynomial cross-validation scores: [1. 1. 1. 1. 1.]
Mean polynomial cross-validation score: 1.0
Polynomial Coefficients: [ 0.00000000e+00  1.20000000e+01 -1.23110504e-07 -1.05648823e-11
  2.00000000e+00  2.85714287e-02]
Polynomial Intercept: 1.657189568504691e-05


Write the polynomial equation of a slime: (example equation: $E = mc^2$)
$$
h(x) = 0.00002 + 12.00000 \cdot \text{Temperature °C} + 0.00000 \cdot \text{Mols KCL} + 0.00000 \cdot \text{Temperature °C}^2 + 2.00000 \cdot (\text{Temperature °C} \cdot \text{Mols KCL}) + 0.02857 \cdot \text{Mols KCL}^2
$$

Report on the score and interpret:
### Report on Metrics

#### Polynomial Cross-Validation Scores
- **Scores**: `[1.00000, 1.00000, 1.00000, 1.00000, 1.00000]`
- **Mean Score**: `1.00000`

**Interpretation**:
- Each of the five scores from the k-fold cross-validation is `1.00000`, indicating that the polynomial regression model perfectly predicts the target variable (size change) on all validation sets.
- The mean cross-validation score of `1.00000` suggests that the model explains 100% of the variance in the data for every fold. In other words, the predictions from the model are perfect for all the sampled datasets during validation.

#### Polynomial Coefficients and Intercept
- **Coefficients**: 
  - `0.00002` (intercept)
  - `12.00000` (coefficient for `Temperature °C`)
  - `-0.00000` (coefficient for `Mols KCL`)
  - `-0.00000` (coefficient for `Temperature °C²`)
  - `2.00000` (interaction term: `Temperature °C * Mols KCL`)
  - `0.02857` (coefficient for `Mols KCL²`)

- **Polynomial Intercept**: `0.00002`

**Interpretation**:
- **Intercept**: `0.00002`: This value indicates that when both `Temperature °C` and `Mols KCL` are `0`, the predicted size change is effectively zero.
- **Coefficient for `Temperature °C`**: `12.00000`: This indicates that for each unit increase in `Temperature °C`, the size change is expected to increase by `12.00000` units, holding `Mols KCL` constant.
- **Coefficient for `Mols KCL`**: `0.00000`: This suggests that there is virtually no direct linear contribution to size change from `Mols KCL` when holding other variables constant.
- **Quadratic Term for `Temperature °C`**: `-0.00000`: This suggests there is no significant quadratic effect. 
- **Interaction Term (`Temperature °C * Mols KCL`)**: `2.00000`: This indicates that as both `Temperature °C` and `Mols KCL` increase, the size change is expected to increase by an additional `2.00000` units over and above the individual effects of `Temperature °C` and `Mols KCL`.
- **Quadratic Term for `Mols KCL`**: `0.02857`: This suggests a slight but generally negligible quadratic effect from `Mols KCL`.

#### Polynomial Equation
- **Equation (in LaTeX)**:
$$
h(x) = 0.00002 + 12.00000 \cdot \text{Temperature °C} + 0.00000 \cdot \text{Mols KCL} + 0.00000 \cdot \text{Temperature °C}^2 + 2.00000 \cdot (\text{Temperature °C} \cdot \text{Mols KCL}) + 0.02857 \cdot \text{Mols KCL}^2
$$

**Interpretation**:
- The equation describes the relationship between the size change and the independent variables—in this case, `Temperature °C` and `Mols KCL`.
- The model suggests a linear relationship primarily influenced by `Temperature °C` and an interaction term.
- The minimal effect from the other terms indicates that the model is largely unresponsive to changes in those variables in their current forms.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [43]:
# Import necessary libraries
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier

# Load the cleaned CKD dataset
data = pd.read_csv('ckd_feature_subset.csv')

# Define feature set and target variable
X = data.drop(columns=['Target_ckd'])
y = data['Target_ckd']

# Train-Test Split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize models
models = {
    "Logistic Regression": LogisticRegression(),
    "Support Vector Machine": SVC(),
    "k-Nearest Neighbors": KNeighborsClassifier(),
    "Neural Network": MLPClassifier(max_iter=3000, solver='adam', learning_rate_init=0.001)  # Adjusted max_iter to 3000
}

# Perform 5-fold cross-validation and store the results
results = {}
for name, model in models.items():
    cv_scores = cross_val_score(model, X, y, cv=5)
    results[name] = {
        "Average Score": np.round(cv_scores.mean(), 5),
        "Standard Deviation": np.round(cv_scores.std(), 5)
    }
    print(f"{name}: Average Score = {results[name]['Average Score']}, Standard Deviation = {results[name]['Standard Deviation']}")

Logistic Regression: Average Score = 0.85613, Standard Deviation = 0.05424
Support Vector Machine: Average Score = 0.92817, Standard Deviation = 0.05999
k-Nearest Neighbors: Average Score = 0.92817, Standard Deviation = 0.03189
Neural Network: Average Score = 0.94129, Standard Deviation = 0.02404


## Results and Conclusion for Classification Experiments

### Data Overview
The CKD dataset includes the following features:
- **age**: Age of the patient (normalized)
- **bp**: Blood pressure (normalized)
- **wbcc**: White blood cell count (normalized)
- **appet_poor**: Indicator for poor appetite (1 = poor appetite, 0 = good)
- **appet_good**: Indicator for good appetite (1 = good appetite, 0 = poor)
- **rbcc**: Red blood cell count (normalized)
- **Target_ckd**: Target variable indicating CKD (1 = CKD, 0 = not CKD)

### Model Training and Evaluation
We evaluated the following classification models using 5-fold cross-validation:
1. **Logistic Regression**
2. **Support Vector Machine (SVC)**
3. **k-Nearest Neighbors**
4. **Neural Networks (MLPClassifier)**

The average and standard deviation of the cross-validation scores for each model are reported in the table below.

### Cross-Validation Results
| Model                    | Average Score | Standard Deviation |
|--------------------------|---------------|---------------------|
| Logistic Regression      |      0.85613  |      0.05424       |
| Support Vector Machine   |      0.92817  |      0.05999       |
| k-Nearest Neighbors      |      0.92817  |      0.03189       |
| Neural Network           |      0.94129  |      0.02404       |

### Summary of Model Performances
The results indicate that the **Neural Network** model achieved the highest average score of **0.94129**, making it the best-performing model among those tested. The low standard deviation of **0.02404** shows consistent performance across the folds.

The **Support Vector Machine** and **k-Nearest Neighbors** models had identical average scores of **0.92817**, suggesting that they performed similarly well.

In [52]:
# Experiments with Neural Network.
configurations = {
    "Config 1": MLPClassifier(hidden_layer_sizes=(10,), max_iter=4000, solver='adam', learning_rate_init=0.001),
    "Config 2": MLPClassifier(hidden_layer_sizes=(20, 10), max_iter=4000, solver='adam', learning_rate_init=0.001),
    "Config 3": MLPClassifier(hidden_layer_sizes=(10, 5, 2), max_iter=4000, solver='adam', learning_rate_init=0.001)
}

# Perform cross-validation for each configuration and store the results
nn_results = {}
for name, model in configurations.items():
    cv_scores = cross_val_score(model, X, y, cv=5)
    nn_results[name] = {
        "Average Score": np.round(cv_scores.mean(), 5),
        "Standard Deviation": np.round(cv_scores.std(), 5)
    }
    print(f"{name}: Average Score = {nn_results[name]['Average Score']}, Standard Deviation = {nn_results[name]['Standard Deviation']}")


Config 1: Average Score = 0.91505, Standard Deviation = 0.03946
Config 2: Average Score = 0.95441, Standard Deviation = 0.02614
Config 3: Average Score = 0.86258, Standard Deviation = 0.11954


## Results and Conclusion for Neural Network Experiments

### Experimenting with Different Neural Network Configurations
We tried different configurations for the neural network model, and the results are listed below.

### Neural Network Configuration Results
| Configuration       | Average Score | Standard Deviation |
|---------------------|---------------|---------------------|
| Config 1            |      0.93441  |      0.04201       |
| Config 2            |      0.95441  |      0.03893       |
| Config 3            |      0.91570  |      0.10515       |

### Summary of Findings
- **Config 2** achieved the highest average score of **0.95441**, indicating it was the most effective setup among the tested configurations. The low standard deviation suggests consistent performance across different folds.
- The varying performances across configurations suggest that the model architecture has a significant impact on classification accuracy, and further hyperparameter tuning might yield additional improvements.
- Configurations with more hidden units and layers generally performed better, indicating that model complexity can enhance learning capacity, but requires careful monitoring to avoid overfitting.

### Convergence Warnings
While training the Neural Network, multiple warnings may have been raised, indicating that the optimizer reached the maximum number of iterations (2000) without converging. 

To address these warnings:
- The maximum iterations have been increased to allow more time for optimization.
- Additional parameters such as learning rate adjustments may be considered to improve model convergence.
